src/cmd/services/m3coordinator/downsample/id_pool_types.go

```diff
@@ -0,0 +1,327 @@
+func (it *encodedTagsIterator) Current() ([]byte, []byte) {
+	tag := it.tagDecoder.Current()
+	return tag.Name.Bytes(), tag.Value.Bytes()
+}
+
+// Err returns any errors encountered.
+func (it *encodedTagsIterator) Err() error {
+	return it.tagDecoder.Err()
+}
+
+// Close closes the iterator.
+func (it *encodedTagsIterator) Close() {
+	it.bytes.Reset(nil)
+	it.tagDecoder.Reset(it.bytes)
+
+	if it.pool != nil {
+		it.pool.Put(it)
+	}
+}
+
+type encodedTagsIteratorPool struct {
+	tagDecoderPool serialize.TagDecoderPool
+	pool           pool.ObjectPool
+}
+
+func newEncodedTagsIteratorPool(
+	tagDecoderPool serialize.TagDecoderPool,
+	opts pool.ObjectPoolOptions,
+) *encodedTagsIteratorPool {
+	return &encodedTagsIteratorPool{
+		tagDecoderPool: tagDecoderPool,
+		pool:           pool.NewObjectPool(opts),
+	}
+}
+
+func (p *encodedTagsIteratorPool) Init() {
+	p.pool.Init(func() interface{} {
+		return newEncodedTagsIterator(p.tagDecoderPool.Get(), p)
+	})
+}
+
+func (p *encodedTagsIteratorPool) Get() *encodedTagsIterator {
+	return p.pool.Get().(*encodedTagsIterator)
+}
+
+func (p *encodedTagsIteratorPool) Put(v *encodedTagsIterator) {
+	p.pool.Put(v)
+}
+
+func isRollupID(
+	sortedTagPairs []byte,
+	iteratorPool *encodedTagsIteratorPool,
+) bool {
+	iter := iteratorPool.Get()
+	iter.Reset(sortedTagPairs)
+
+	tagValue, ok := iter.TagValue(rollupTagName)
+	iter.Close()
+
+	return ok && bytes.Equal(tagValue, rollupTagValue)
+}
+
+// rollupIDProvider is a constructor for rollup IDs, it can be pooled to avoid
+// requiring allocation every time we need to construct a rollup ID.
+// When used as a ident.TagIterator for the call to serialize.TagEncoder Encode
+// method, it will return the rollup tag in the correct alphabetical order
+// when progressing through the existing tags.
+type rollupIDProvider struct {
+	index          int
+	tagPairs       []id.TagPair
+	rollupTagIndex int
+
+	tagEncoder serialize.TagEncoder
+	pool       *rollupIDProviderPool
+}
+
+func newRollupIDProvider(
+	tagEncoder serialize.TagEncoder,
+	pool *rollupIDProviderPool,
+) *rollupIDProvider {
+	return &rollupIDProvider{
+		tagEncoder: tagEncoder,
+		pool:       pool,
+	}
+}
+
+func (p *rollupIDProvider) provide(
+	tagPairs []id.TagPair,
+) ([]byte, error) {
+	p.reset(tagPairs)
+	p.tagEncoder.Reset()
+	if err := p.tagEncoder.Encode(p); err != nil {
+		return nil, err
+	}
+	data, ok := p.tagEncoder.Data()
+	if !ok {
+		return nil, fmt.Errorf("unable to access encoded tags: ok=%v", ok)
+	}
+	// Need to return a copy
+	id := append([]byte(nil), data.Bytes()...)
+	// Reset after computing
+	p.reset(nil)
+	return id, nil
+}
+
+func (p *rollupIDProvider) reset(
+	tagPairs []id.TagPair,
+) {
+	p.index = -1
+	p.tagPairs = tagPairs
+	p.rollupTagIndex = -1
+	for idx, pair := range tagPairs {
+		if bytes.Compare(rollupTagName, pair.Name) < 0 {
+			p.rollupTagIndex = idx
+			break
+		}
+	}
+	if p.rollupTagIndex == -1 {
+		p.rollupTagIndex = len(p.tagPairs)
```

Comment:
Because its not part of the tag pairs, the rollup ID provider actually wraps tag pairs and when enumerated gives you back the rollup tag at the right spot.  It's to save re-allocing tag pairs each time when you need to insert the tag, you can just get a pooled rollup ID provider, set the tag pairs and then give it to another component that can iterate them and get back the rollup tag at the right position when enumerating the tags.